File: gptcomet/llms/claude.py
from typing import Any, Optional

from .base import BaseLLM


class ClaudeLLM(BaseLLM):
    def __init__(self, config: dict[str, Any]):
        super().__init__(config)

        self.api_base = config.get("api_base") or "https://api.anthropic.com/v1"
        self.model = config.get("model") or "claude-3-5-sonnet"
        self.completion_path = config.get("completion_path") or "messages"
        self.answer_path = config.get("answer_path") or "content.0.text"
        self.anthropic_version = config.get("anthropic-version", "2023-06-01")
        self.top_k = config.get("top_k")

    def build_headers(self):
        headers = {
            "x-api-key": self.api_key,
            "Content-Type": "application/json",
            "anthropic-version": self.anthropic_version,
        }
        return headers

    def format_messages(
        self, message: str, history: Optional[list[dict[str, str]]] = None
    ) -> dict[str, Any]:
        payload = {
            "model": self.model,
            "prompt": message,
            "max_tokens": self.max_tokens,
        }
        if self.top_k:
            payload["top_k"] = self.top_k

        if self.temperature is not None:
            payload["temperature"] = float(self.temperature)

        if self.top_p is not None:
            payload["top_p"] = float(self.top_p)

        return payload

    def get_usage(self, data: dict[str, Any]) -> Optional[str]:
        usage = data.get("usage")
        if not usage:
            return None
        return (
            f"Token usage: input tokens: {usage.get('input_tokens')}, "
            f"output tokens: {usage.get('output_tokens')}, "
        )

    @classmethod
    def get_required_config(cls) -> dict[str, tuple[str, str]]:
        return {
            "api_base": ("https://api.anthropic.com/v1", "Enter Anthropic API Base URL"),
            "model": ("claude-3-5-sonnet", "Enter model name"),
            "api_key": ("", "Enter API key"),
            "max_tokens": ("1024", "Enter max tokens"),
            "anthropic-version": ("2023-06-01", "Enter Anthropic API version"),
        }

File: internal/llm/claude.go
package llm

import (
	"context"
	"fmt"
	"net/http"
	"strings"

	"github.com/tidwall/gjson"

	"github.com/belingud/go-gptcomet/pkg/config"
	"github.com/belingud/go-gptcomet/pkg/types"
)

// ClaudeLLM is the Claude LLM provider implementation
type ClaudeLLM struct {
	*BaseLLM
}

// NewClaudeLLM creates a new ClaudeLLM
func NewClaudeLLM(config *types.ClientConfig) *ClaudeLLM {
	if config.APIBase == "" {
		config.APIBase = "https://api.anthropic.com/v1"
	}
	if config.Model == "" {
		config.Model = "claude-3-sonnet"
	}
	if config.CompletionPath == "" {
		config.CompletionPath = "messages"
	}
	if config.AnswerPath == "" {
		config.AnswerPath = "content.0.text"
	}
	if config.AnthropicVersion == "" {
		config.AnthropicVersion = "2024-01-01" // 使用最新的稳定版本
	}

	return &ClaudeLLM{
		BaseLLM: NewBaseLLM(config),
	}
}

func (c *ClaudeLLM) Name() string {
	return "Claude"
}

// GetRequiredConfig returns provider-specific configuration requirements
func (c *ClaudeLLM) GetRequiredConfig() map[string]config.ConfigRequirement {
	return map[string]config.ConfigRequirement{
		"api_base": {
			DefaultValue:  "https://api.anthropic.com/v1",
			PromptMessage: "Enter Claude API base",
		},
		"model": {
			DefaultValue:  "claude-3-sonnet",
			PromptMessage: "Enter model name",
		},
		"api_key": {
			DefaultValue:  "",
			PromptMessage: "Enter API key",
		},
		"anthropic_version": {
			DefaultValue:  "2024-01-01",
			PromptMessage: "Enter Anthropic API version",
		},
		"max_tokens": {
			DefaultValue:  "1024",
			PromptMessage: "Enter max tokens",
		},
	}
}

// FormatMessages formats messages for Claude API
func (c *ClaudeLLM) FormatMessages(message string, history []types.Message) (interface{}, error) {
	messages := make([]map[string]interface{}, 0, len(history)+1)
	messages = append(messages, map[string]interface{}{
		"role":    "user",
		"content": message,
	})

	payload := map[string]interface{}{
		"model":             c.Config.Model,
		"messages":          messages,
		"max_tokens":        c.Config.MaxTokens,
		"temperature":       c.Config.Temperature,
		"top_p":             c.Config.TopP,
		"frequency_penalty": c.Config.FrequencyPenalty,
		"presence_penalty":  c.Config.PresencePenalty,
	}

	return payload, nil
}

// BuildURL builds the API URL
func (c *ClaudeLLM) BuildURL() string {
	return fmt.Sprintf("%s/%s", strings.TrimSuffix(c.Config.APIBase, "/"), strings.TrimPrefix(c.Config.CompletionPath, "/"))
}

// BuildHeaders builds request headers
func (c *ClaudeLLM) BuildHeaders() map[string]string {
	headers := map[string]string{
		"Content-Type":      "application/json",
		"anthropic-version": c.Config.AnthropicVersion,
		"x-api-key":         c.Config.APIKey,
	}
	for k, v := range c.Config.ExtraHeaders {
		headers[k] = v
	}
	return headers
}

// ParseResponse parses the response from the API
func (c *ClaudeLLM) ParseResponse(response []byte) (string, error) {
	text := gjson.GetBytes(response, c.Config.AnswerPath).String()
	if strings.HasPrefix(text, "```") && strings.HasSuffix(text, "```") {
		text = strings.TrimPrefix(text, "```")
		text = strings.TrimSuffix(text, "```")
	}
	return strings.TrimSpace(text), nil
}

// GetUsage returns usage information for the provider
func (c *ClaudeLLM) GetUsage(data []byte) (string, error) {
	usage := gjson.GetBytes(data, "usage")
	if !usage.Exists() {
		return "", nil
	}

	return fmt.Sprintf(
		"Token usage> input tokens: %d, output tokens: %d",
		usage.Get("input_tokens").Int(),
		usage.Get("output_tokens").Int(),
	), nil
}

// MakeRequest makes a request to the API
func (c *ClaudeLLM) MakeRequest(ctx context.Context, client *http.Client, message string, history []types.Message) (string, error) {
	return c.BaseLLM.MakeRequest(ctx, client, c, message, history)
}
